### DISCUSSION TITLE
Can't import CSV's

### DISCUSSION THREAD

**Discussion by ghost** [netbox-community/netbox#10326] [Q&A]
I am attempting to export and import items from CSV's. Through trial and error I have determined that when NetBox exports CSV's the field names are similar, but not what NetBox accepts during import.
"Name" instead of "name"
"Content Types" instead of "content_types"

The field names are also case sensitive, which isn't documented anywhere btw.
Anyway, I've managed to manually clean a CSV export and get it all the way to the very last column. In my particular example I'm dealing with the 'custom fields' section.
There is a column called 'ui_visibility'. When doing an export, the output in the CSV is 'Read/Write'. However, during the import it fails saying that 'Read/Write' is not a valid option. Where are the options documented? For type, content_types, and object_types there is a question mark at the end of the row that gives you a popup of all the possible options. But this doesn't exist for 'ui_visibility'

I've tried 'read/write' all lowercase, I've tried 'readwrite'. I'm at a loss here and hopefully someone can assist.

**Comment by jsenecal** [comment]
Hi @cyr0nk0r, we appreciate that you edited your original post but remember that everything that you put on the internet stays on the internet, forever.

I understand from your other few opened issues and questions (https://github.com/netbox-community/netbox/issues/10321, https://github.com/netbox-community/netbox/discussions/10323) that you are having issues with the asymmetry related to the export and import formats that NetBox allows. 

The format is indeed asymmetrical **by design** as the export format was intended for Humans to read/use whilst the import format is intended for NetBox to use and thus, needs to be machine-readable.

Everything pertaining to the import format is however very well documented on the import page of each models, for instance https://demo.netbox.dev/dcim/devices/import/ (use admin/admin to login if you haven't already). The supported fields and how the columns should be identified in the csv file are right in there, and the `accessor` describes which field the import tool is looking for when parsing each row.

If you feel that [the documentation](https://docs.netbox.dev/en/stable/getting-started/populating-data/#bulk-import-csvyaml) is inappropriate and does not cover accurately the process of doing Bulk Imports, then I encourage you to open a documentation issue describing accurately what you would like in there and your thought process.

In regards of "feeding back to NetBox its exported data", would you mind sharing an example of use-case? Maybe there is a better way (aka you're holding it wrong) or maybe you're right and we should do something about this user experience related issue.

In any case, remember to treat the NetBox Community with due respect as failure to do so will not be tolerated. We are a group of enthusiasts and most of us are contributing to NetBox on our own personal time, so please be mindful of that. I recommend that you read through [GitHub's best practices](https://docs.github.com/en/site-policy/github-terms/github-community-code-of-conduct#best-practices-for-maintaining-a-strong-community) on the topic.

**Comment by kkthxbye-code** [comment]
> There is a column called 'ui_visibility'. When doing an export, the output in the CSV is 'Read/Write'. However, during the import it fails saying that 'Read/Write' is not a valid option.

read-write, read-only, hidden

**Reply by jsenecal** [reply]
I opened https://github.com/netbox-community/netbox/issues/10333 to track the lack of documentation button for that field.

**Comment by ghost** [comment]
Team, I understand you can see comments before they were edited. And I hear you regarding due respect. I would say that others on your team also need to treat the community with due respect. My frustration is not directed at either you Jonathan, or kkthxbye-code, since both of you actually took the time to either attempt to help, or provide the information that is tripping me up.

What I don't appreciate is submitting a discussion or feature request or bug report and having them closed with what is essentially a RTFM from Jeremy. I will reiterate that I do not understand why he has such an adversarial attitude towards people just trying to learn.

We are not all Linux experts, we are not all programmers. And when my submitted bug request was closed out with a curt response I then moved things over to discussions so I could hopefully better understand what my problem is, which Jeremy then again just closed with another 'your question was answered'. No actually, it wasn't answered. Since neither is it documented anywhere that field names are case-sensitive, nor is the ui-visibility column's available options documented anywhere.

**Reply by jcollie** [reply]
> I will reiterate that I do not understand why he has such an adversarial attitude towards people just trying to learn.

I find that in maintainers of large open-source projects that this is usually a learned defense mechanism.

**Reply by jsenecal** [reply]
> Since neither is it documented anywhere that field names are case-sensitive, nor is the ui-visibility column's available options documented anywhere.

The field names and Bulk Import process are indeed properly documented as indicated in my previous answer, if you don't agree with that statement "then I encourage you to open a documentation issue describing accurately what you would like in there and your thought process". 
In regards of the missing help text for `ui_visibility`, that was indeed an omission or a bug which Jeremy himself addressed in https://github.com/netbox-community/netbox/pull/10335 after being flagged as such. Unfortunately only this last discussion had a reference or mention to this omission.

Nothing here is personal, we try our best to help and sometimes the perceived tone of the issue author tips the balance in a direction or the other. Now, from my personal point of view, Jeremy did spend the time to reply and give you information and yet you were rude in your subsequent interactions.

Again, the community does the best it can with the information it has...

**Reply by peterbaumert** [reply]
> > I will reiterate that I do not understand why he has such an adversarial attitude towards people just trying to learn.
> 
> 
> 
> I find that in maintainers of large open-source projects that this is usually a learned defense mechanism.
> 
> 

It is not a big secret that he hasn't the best community skills though. Of the few people i have talked to about netbox this is the one common thing everyone brings up 🤷‍♂️

Maybe it would help at least considering that there might be space to improve.

**Comment by ghost** [comment]
If you say the field is documented properly called 'name', but 'Name' doesn't work and there is no mention that 'Name' will fail or that it MUST be 'name' and only 'name' as being properly documented then I guess we'll have to agree to disagree.

If you seriously don't see the problem with a field being case sensitive and not making ANY mention that the field name must be case sensitive as proper documentation then there isn't anything more I can say here.

**Reply by jsenecal** [reply]
Yes, I _seriously_ don't consider this a problem but you have all the rights to consider this a usability issue.

Would this work for you ?
![image](https://user-images.githubusercontent.com/1564902/189719834-478d222e-5451-474b-ae06-f3c62e0445a9.png)

**Reply by ghost** [reply]
It could be there where you have it, it could be at the bottom of the page, it could be in the documentation. I mean, it could be anywhere .. as long as it's somewhere.

Personally I think your screenshot is the best scenario, and it would go a long way to help users.

**Reply by jsenecal** [reply]
Glad that we got something that covers your issue.

Can you now please open a [Feature Request](https://github.com/netbox-community/netbox/issues/new?assignees=&labels=type%3A+feature&template=feature_request.yaml) suggesting those changes and describe how they could help newcomers to interact with NetBox ? You can use my screenshot and even refer to this discussion if you'd like.

Thanks!

**Reply by ghost** [reply]
Will do. Thanks

**Reply by ghost** [reply]
> Glad that we got something that covers your issue.
> 
> Can you now please open a [Feature Request](https://github.com/netbox-community/netbox/issues/new?assignees=&labels=type%3A+feature&template=feature_request.yaml) suggesting those changes and describe how they could help newcomers to interact with NetBox ? You can use my screenshot and even refer to this discussion if you'd like.
> 
> Thanks!

Exactly as I predicted. 
https://github.com/netbox-community/netbox/issues/10344#issuecomment-1248165701
Out of spite Jeremy just closes the request and continues to reiterate that things are properly documented.

**Reply by peterbaumert** [reply]
q.e.d. :)
